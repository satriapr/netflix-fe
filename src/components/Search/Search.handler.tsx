import { isEmpty } from 'ramda'
import { useEffect, useState } from 'react'

import Constant from '../../constants/Constant'
import { IMovieRes } from '../../models'
import movieService from '../../services/movie.service'

const Handler = () => {
  const [val, setVal] = useState<string>('')
  const [movies, setMovies] = useState<IMovieRes[]>([])
  const [showRes, setShowRes] = useState<boolean>(false)

  const getMovies = async (input: string) => {
    const response = await movieService.getSearchedMovies(input)
    setMovies(response)
    setShowRes(true)
  }

  const handleChange = e => setVal(e.target.value)

  // Debounce. Only call API after last key of user input
  useEffect(() => {
    const input = val.toLowerCase()
    if (!isEmpty(input) && input.length >= Constant.MINIMUM_SEARCH_KEYWORD) {
      const timeout = setTimeout(() => {
        getMovies(input)
      }, Constant.SEARCH_DELAY)

      return () => clearTimeout(timeout)
    }

    setMovies([])
    setShowRes(false)
  }, [val])

  const hideRes = () => {
    setShowRes(false)
    setVal('')
  }

  return { val, movies, showRes, handleChange, hideRes }
}

export default Handler
