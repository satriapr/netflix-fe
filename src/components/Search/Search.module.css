
.searchInput {
  background: none;
  color: #fff;
  border: 1px solid #fff;
  outline: none;
  padding: 5px;
  padding-left: 15px;
  width: 250px;
  height: 20px;
  font-weight: 600;
  margin-top: 5px;
  position: relative;
  font-size: 14px;
}

.searchValues {
  background: #0d0d0d;
  width: 268px;
  border: 2px solid #fff;
  border-top: 0;
  border-top-left-radius: 0;
  border-top-right-radius: 0;

  position: absolute;
  padding: 0;
  z-index: 1;
}

.searchValues ul {
  margin: 0;
  padding: 10px;
  list-style: none;
}

.searchValues ul li {
  margin: 8px;
}

.searchValues a {
  text-decoration: none;
  color: #FFF;
}

.searchValues a:hover {
  text-decoration: underline;
  font-weight: 600;
}

@media only screen and (max-width: 690px) {
  .searchInput {
    width: 95%;
  }

  .searchValues {
    width: 87%;
  }
}