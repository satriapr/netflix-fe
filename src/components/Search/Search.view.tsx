import React from 'react'
import { Link } from 'react-router-dom'

import Constant from '../../constants/Constant'
import searchHandler from './Search.handler'
import styles from './Search.module.css'

const Search = () => {
  const handler = searchHandler()
  const { val, movies, handleChange, showRes, hideRes } = handler

  const movieList = movies.length ? (
    movies.map(movie => {
      return (
        <li key={movie.imdbID}>
          <Link to={`/${movie.imdbID}`} onClick={hideRes}>
            {movie.Title}
          </Link>
        </li>
      )
    })
  ) : (
    <li>{Constant.MOVIE_NOT_FOUND}</li>
  )

  return (
    <React.Fragment>
      <input
        type="text"
        name="searchVal"
        onChange={handleChange}
        className={styles.searchInput}
        placeholder={Constant.SEARCH_PLACEHOLDER}
        value={val}
      />
      {showRes && val.length >= Constant.MINIMUM_SEARCH_KEYWORD && (
        <div className={styles.searchValues}>
          <ul>{movieList}</ul>
        </div>
      )}
    </React.Fragment>
  )
}

export default Search
