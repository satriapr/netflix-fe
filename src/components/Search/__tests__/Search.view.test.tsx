import { testSnapshots } from '../../../utils/test.util'
import Search from '../Search.view'

describe('Search Component', () => {
  testSnapshots(Search, [
    {
      description: 'Search props',
      props: {}
    }
  ])
})
