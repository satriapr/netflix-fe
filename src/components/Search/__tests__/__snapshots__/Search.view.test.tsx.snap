// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component snapshots Search props 1`] = `
<Fragment>
  <input
    name="searchVal"
    onChange={[Function]}
    placeholder="Titles"
    type="text"
    value=""
  />
</Fragment>
`;
